mainapp/src/Admin.tsx

```diff
@@ -1,30 +1,389 @@
+        putUsers('/users', [...users, user])
+            .then((person) => updateModel({ type: 'ServerRespondedWithUsers', payload: success(person) }))
+            .then(() => props.setModal(false))
+            .then(() => setNewUser(newUser(ulid())))
+            .catch((err) => updateModel({ type: 'ServerRespondedWithUsers', payload: failure(err.msg) }))
+    }
+
+
+    return (
+        <>
+            <Box mb={4}><Button color="success" variant="contained" onClick={() => props.setModal(true)}>Add user</Button></Box>
+            {//newFunction(props, setNewUser, user, profiles, saveUser, deletedUser)
+            }
+            <UserForm
+                modal={props.modal}
+                updateModel={props.updateModel}
+                setModal={props.setModal}
+                setNewUser={setNewUser}
+                user={user}
+                profiles={profiles}
+                saveUser={saveUser}
+                deletedUser={deleteUser(users, user, updateModel, props.setModal)} />
+        </>
+    )
+}
+
+type UserProps = { user: User }
+
+const User: React.FC<UserProps> = ({ user }) => {
+
+    const { model, updateModel } = useModel();
+
+    const [localUser, setNewUser] = React.useState(user);
+
+    const [isModalOpen, setModal] = React.useState(false);
+
+    const restoredUsers = restoreUsers(updateModel, setModal)
+
+    const handleOpen = () => setModal(true);
+
+
+    const users: User[] = SRD.unwrap([], identity, model.users)
+
+    const profiles: string[] = SRD.unwrap([], identity, model.profiles)
+
+
+    return (<>
+
+        <Button variant="outlined" size="medium" onClick={handleOpen}>{`Edit ${user.login}`}</Button>
+        <UserForm
+            modal={isModalOpen}
+            updateModel={updateModel}
+            setModal={setModal}
+            setNewUser={setNewUser}
+            user={localUser}
+            profiles={profiles}
+            saveUser={saveUser(updateModel, users, user, localUser, setModal)}
+            deletedUser={deleteUser(users, user, updateModel, setModal)} />
+
+    </>)
+};
+
+type UserFormProps = {
+    modal: boolean,
+    updateModel: React.Dispatch<Msg>,
+    setModal: React.Dispatch<React.SetStateAction<boolean>>,
+    setNewUser: React.Dispatch<React.SetStateAction<User>>,
+    user: User,
+    profiles: string[],
+    saveUser: () => void,
+    deletedUser: () => void
+}
+
+const UserForm: React.FC<UserFormProps> = ({ modal, setModal, updateModel, setNewUser, user, profiles, saveUser, deletedUser }) => {
+    return <Modal open={modal}
+        onClose={restoreUsers(updateModel, setModal)}
+        aria-labelledby="modal-modal-title"
+        aria-describedby="modal-modal-description">
+        <Box sx={style}>
+            <Stack spacing={4}>
+                <h2>{`Edit ${user.login}`}</h2>
+                <TextField fullWidth onChange={(event) => setNewUser({ ...user, login: event.target.value })}
+                    //@ts-ignore
```

Comment:
why do you nee a ts-ignore here ?

Reply:
It's removed ! I needed it at some point because typescript type inference went weird. Looks it's resolved!